### ISSUE TITLE
ENH: Should there be an `rng.clone()` or similar?

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#24086]
### Proposed new feature or change:

Discussing with @betatim, sklearn sometimes would like to re-use the same random number state (e.g. for splitting data the same way, but potentially many times).

I am not sure about whether that is a good idea, but I assume there is a need for this.  So this is to track/get feedback if we should add a `.clone()` or `.copy()` method to the `rng`/`bit_generator`.

Right now, the best pattern I could think of is to basically do:
```
def __init__(self, rng=None)
    # maybe spawn a new one to have one for ourself
    # (not sure how it exactly looks like)
    self._blueprint_rng, = new_rng(rng) or rng.spawn(1)

def method(self):
    rng = copy.deepcopy(self._blueprint_rng)
    # use rng
```
The reason is that we have to work with a copy because otherwise threading will be broken.  `deepcopy` works, and it might be nice to implement `__deepcopy__` to reduce the overhead a bit (I think this could be <1us rather than >10us easily).

But the other question is whether `copy.deepcopy()` is an obvious enough solution to begin with, or whether it wouldn't be better to have an explicit method to make this easy?

Or maybe I am missing a nicer pattern to have such a "rewinding", rng?

**Comment by rkern** [comment]
One improvement to the pattern would be to avoid having a live, operational `Generator` be the static, stored object. That's mildly error-prone, because if anyone calls any methods on it in the interim, you'll break the desired semantics. Rather, we might consider adding a function (not a method) that kind of works like `default_rng()`, except that it returns an opaque object that holds a `SeedSequence` and the `BitGenerator` type and will construct a fresh `Generator` from those two on request. When given a static seed like an integer (or a sequence of ints), then it just returns `SeedSequence(seed)` (and I think it should pass through a `SeedSequence` if one is passed, but I could be convinced that it should spawn). But if given a `Generator`, it will reach down and spawn from the underlying `SeedSequence` and use that spawned `SeedSequence`, which is now owned by this object. When you kick off the computation, you call this opaque object to get a fresh `Generator` and go.

The main problem with this scheme is mostly finding a name that's not super-confusing.

```python
from copy import deepcopy
from functools import partial

import numpy as np


def reusable_rng_factory(rng=None):
    if isinstance(rng, np.random.Generator):
        # Grab the BitGenerator and fall through to the next if clause
        rng = rng._bit_generator
    if isinstance(rng, np.random.BitGenerator):
        seed_seq = rng._seed_seq.spawn(1)[0]
        bg_type = type(rng._bit_generator)
        # Why the deepcopy? SeedSequence has some state concerning the number of spawns it has done.
        # To get the same starting state each time, we need 
        restorer = lambda seed: np.random.Generator(bg_type(deepcopy(seed)))
    else:
        seed_seq = np.random.SeedSequence(rng)
        restorer = lambda seed: np.random.default_rng(deepcopy(seed))
    # Obviously, this probably should be a real type with a name and docstrings, but this gets across the idea.
    # Basically opaque, with just a __call__ method for the API.
    return partial(restorer, seed_seq)


class MyFancyComputation:
    def __init__(self, rng=None):
        self._rng_factory = reusable_rng_factory(rng)

    def method(self):
        rng = self._rng_factory()
        x = rng.standard_normal(1000)
        ...
````

I'm mostly not in favor of having anything like rewinding a live `Generator/BitGenerator` to a previous state. There's already too much statefulness running around to keep track of.

**Comment by seberg** [comment]
Right, if we have this as API in NumPy that is nice in the sense that we take care of giving the user a way to fetch identical rng instances without exposing their "state" (even from the initial one, since we ensure the `.spawn()`).  (Which is something `.clone()` would not do!)

One reason other reason I brought it up is discoverability, but I am OK with this being not super discoverable:  SPEC 7 should mention it if sklearn starts to use it, I expect.  That seems good enough.

I would be happy if we include this into NumPy, maybe "identical" is not a bad name fragment?  If there is a use-case for exposing an `rng_factory()` that spawns new ones, this could be a kwarg option also.

**Comment by betatim** [comment]
One use-case (hypothetical because the whole use in scikit-learn is hypothetical for now, after all no one has started writing any code) is that users will use `default_rng()` and pass that in to our estimators. So maybe the thing to add to numpy is something that can take a "live" generator and return a "not live" one. The "not live" one could be used to create new ones but would raise an error when you call something like `.random()` (or other things that modify its state).

Yet another twist (maybe beyond this issue?) is that in scikit-learn we currently adhere to the principle that `__init__` doesn't do any work ([more details](https://scikit-learn.org/dev/developers/develop.html#instantiation)). But this might be a particularity of scikit-learn and in a more general world there are still people who have a need for turning a "live" generator into a "not live" one?

**Comment by betatim** [comment]
Taking two more seconds to read the proposed code of the RNG factory, this might be exactly what I described in my comment. But, (see comment on naming) when I first read it the word "Factory" made me think that a factory is something that makes things, so you wouldn't go from the thing back to a factory.

**Comment by ogrisel** [comment]
I think the line `bg_type = type(rng._bit_generator)` should be replaced by `bg_type = type(rng)`.

But beyond this, the line `seed_seq = rng._seed_seq.spawn(1)[0]` would mutate the state of the rng instance passed as argument to `reusable_rng_factory`.

Any efficient implementation of `__deepcopy__` would indeed make it possible for a library to derive generators without mutating the ancestor and therefore recover for that snapshot.

I think it would be even better to provide an explicitly stateless RNG API similar to what Jax does: the users of that API would be guaranteed to never make a mistake that would mutate an "rng key" passed as an argument to their function/method. To make it progress, they would have to "split" it first.

**Comment by rkern** [comment]
Yeah, but it follows the general conventions of naming functions. When functions are nouns, they tend to be the things that are returned by the function. `default_rng()` returns an RNG, `mean()` returns an arithmetic mean, `array()` returns an array, `reusable_rng_factory` returns a factory. But noted.

**Comment by rkern** [comment]
> I think the line `bg_type = type(rng._bit_generator)` should be replaced by `bg_type = type(rng)`.

Yes, copypasta. Fixed.

> But beyond this, the line `seed_seq = rng._seed_seq.spawn(1)[0]` would mutate the state of the rng instance passed as argument to `reusable_rng_factory`.

Yes. That's fine and correct behavior for this pattern. What this function needs to guarantee is that the returned factory object returns a `Generator` with the same state on each call and that that `Generator` is different from all other `Generator`s not created by that particular factory object.

> Any efficient implementation of `__deepcopy__` would indeed make it possible for a library to derive generators without mutating the ancestor and therefore recover for that snapshot.
> 
> I think it would be even better to provide an explicitly stateless RNG API similar to what [Jax does](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#jax-prng): the users of that API would be guaranteed to never make a mistake that would mutate an "rng key" passed as an argument to their function/method. To make it progress, they would have to "split" it first.

My thoughts on the appropriateness of that style of PRNG for numpy [can be found here](https://github.com/data-apis/array-api/issues/431#issuecomment-1147804063). Ultimately, I think you are just trading types of user errors (on top of the lowered safety margins of splitting algorithms). The places where you want the PRNG to progress are vastly more frequent than ones where you don't. So you are creating a situation where it's more likely that a user reuses a key accidentally. Either way, we are long past the discussion window for `np.random`.

**Comment by rkern** [comment]
> Any efficient implementation of `__deepcopy__` would indeed make it possible for a library to derive generators without mutating the ancestor and therefore recover for that snapshot.

Forgot to respond to this. You do _not_ want to clone the `Generator` that is passed in. You must spawn a `Generator` from that `Generator` before you can clone a copy from it. The rule is [*Only clone it if you own it*](https://github.com/scikit-learn/scikit-learn/issues/16988#issuecomment-1518037853). Otherwise, the original passed-in `Generator` will continue on _outside_ of the CV splitter (in this example) and generate correlated values with those used _inside_ of the CV splitter, and that is not what you want. The only things you want to be correlated are the things you deliberately designed to be correlated: multiple calls to the splitter.

**Comment by seberg** [comment]
Agreed, I hate the idea that a user could get the same stream of randomness twice without being overly explicit about it. Passing in an rng instance isn't that:  Normally, if I do that I expect you to use it directly if anything as in `rng.uniform()` which *also* modifies state.

Modifying state is a feature, not a bug.  If we were in JAX world, the user would have to spawn for us.  But even if the user expects JAX for whatever reason: it doesn't seem like it would matter as the additional spawn is in practice irrelevant.

**Comment by ogrisel** [comment]
Ok thanks for your replies. Those are interesting considerations.

**Comment by seberg** [comment]
We have this now.  One thing that came up once is that we could allow `.clone()` without any numbr to return a single new child rathre than a tuple

**Comment by ricardoV94** [comment]
> We have this now.

@seberg what do you mean. Is there a `.clone` method for `Generator` objects. This is one thing that we are facing in the transition to numpy 2.x at PyTensor: https://github.com/pymc-devs/pytensor/pull/689#issuecomment-2626917728

As a non-eager computational library we are in the JAX-camp where we want non-mutability by default (and later specialize to mutate, as copying is obviously slow). We use to simply copy the rng before using it, now I guess we can use `deepcopy`. Would still be nice if deepcopy had low overhead, or if there was a `.clone` method. I can't find it in the documentation and there was no PR linked to this issue?

**Comment by seberg** [comment]
I'll re-open, because maybe I got confused with `.spawn()` and `.clone()`.

That said, @rkern did give a pattern to get a *repeatable* `rng` instance.  And it is not clear to me yet that you need to repeatability!

If you do *not* need repeatability, then the right thing is to use `.spawn()[0]` (and even if you need it, you need to use spawning first).

IIRC, JAX asks you to do `seed, state = split_state(state)` while NumPy just does `new_rng, = rng.split()`.
And in JAX you should (almost) always overwrite the original `state` since you do not want to use the same state twice accidentally!
There are rare cases where you want to use the same state twice and the "repeatable" pattern is an API Design that would fit `sklearn`.

In your case, rather than the `repeatable` design based on `RandomGenerator`/`rng`, it may instead make sense to work with `seed = np.random.SeedSequence()` instead.  Since `rng = np.random.default_rng(seed)` will always give the same stream.
(Of course multiple `seed.spawn(1)` will keep giving you unique children, but that is a feature, not a bug.  Everyone who needs repeatability has to hold on to "their" `seed` to reinstantiate the `rng`.)

**Comment by rkern** [comment]
Given that you want to consume regular `Generator`s and not invent your own objects, then yes `deepcopy()` is probably what you want. I might have some concerns about mixing pytensor code with plain old numpy, but once you're inside the pytensor world and stay there, this pattern is probably fine. Consider recommending or forcing a `spawn()` when transitioning from plain-old-numpy code to pytensor-owned calculations, but I haven't spent the time to think through the whole pytensor design yet.

I don't think the ceepcopy can be made significantly faster given the constraints (`BitGenerator`s and `ISeedSequence`s can be contributed by 3rd party code), but you're welcome to prototype something to propose with supporting benchmarks.

**Comment by ricardoV94** [comment]
> Consider recommending or forcing a spawn() when transitioning from plain-old-numpy code to pytensor-owned calculations, but I haven't spent the time to think through the whole pytensor design yet.

We do spawn in the more-user friendly API, and we mention that in the low-level doc: https://pytensor.readthedocs.io/en/latest/tutorial/prng.html#creating-multiple-rng-variables (if that's what you meant)

> I don't think the ceepcopy can be made significantly faster given the constraints (BitGenerators and ISeedSequences can be contributed by 3rd party code), but you're welcome to prototype something to propose with supporting benchmarks.

Okay, in that case the issue can be closed. I brought it up because of this part of the top-comment:

> The reason is that we have to work with a copy because otherwise threading will be broken. deepcopy works, and it might be nice to implement __deepcopy__ to reduce the overhead a bit (I think this could be <1us rather than >10us easily).

**Comment by ricardoV94** [comment]
Back to this. Deepcopy is a bit too wasteful? IIRC (at least for PCG64()) we just need to create a new bit_generator object and deepcopy the dictionary state?

My implementation below is 2x faster than deecopy, but it's still wasteful. Most work comes from initializing a new bit_generator to then override its state.
```python
from copy import deepcopy
from numpy.random import default_rng, PCG64

def custom_deepcopy(rng):
    orig_bitgen = rng.bit_generator
    # Custom seed faster than `None`, but still slow
    new_bitgen = type(orig_bitgen)(0)
    # Probably could have a specialized copy, just need shallow copy of state and state["state"] in case of PCG64
    new_bitgen.state = deepcopy(orig_bitgen.state)
    return default_rng(new_bitgen)

rng = default_rng()
assert deepcopy(rng).bit_generator.state == custom_deepcopy(rng).bit_generator.state
%timeit deepcopy(rng)  # 66.6 μs ± 1.96 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit custom_deepcopy(rng)  # 32.6 μs ± 1.47 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Would it be possible to allow create a new PCG64 from a given state, instead of always requiring a seed? Something like `PCG64.from_state(dict)`?

I'm assuming that the initializing the state is the slow part. This could be plainly wrong ofc

PS: It's obviously fine to decline
PPS: This could also be used for `.jumped`, which also creates a new Generator and then overrides its state
PPPS: We PCG64.from_state is basically `PCG64(SeedSequence)`

Here is a correct? impl that's still faster

```python
from copy import deepcopy
from numpy.random import default_rng, Generator

def custom_deepcopy(rng):
    old_bitgen = rng.bit_generator
    old_seed_seq = old_bitgen.seed_seq
    new_seed_seq = deepcopy(old_seed_seq)
    new_bitgen = type(old_bitgen)(new_seed_seq)
    new_bitgen.state = old_bitgen.state  # state is a method that returns a fresh dict
    return Generator(new_bitgen)

rng = default_rng()
dp = deepcopy(rng).bit_generator
cc = custom_deepcopy(rng).bit_generator
assert dp.state == cc.state
assert dp.seed_seq.state == cc.seed_seq.state
%timeit deepcopy(rng) # 66.6 μs ± 1.96 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit custom_deepcopy(rng) # 27.6 μs ± 701 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

**Comment by seberg** [comment]
> Deepcopy is a bit too wasteful?

@ricardoV94 one can implement `__deepcopy__` to side-step whatever the default deepcopy implemenetation is doing that makes it slow, and if you do it for both the bit_generator and the generator presumably you can make it fast (enough).
That may be even faster since it probably allows you to side-step the state dict creation.

IIRC I was initially wondering if we should have something more visible than a deepcopy, implementing `__deepcopy__` to make it faster isn't an API issue though and can just be done.

**Comment by ricardoV94** [comment]
Hmm `BitGenerator` state is no longer just state, it also references mutable `self._seed_seq`...

**Comment by ricardoV94** [comment]
Do I understand this would be done in https://github.com/numpy/numpy/blob/cfda63ddd2a7cf06b77a09d5834f053c092bd85a/numpy/random/bit_generator.pyx#L492

I assume the inability to set `bit_generator._seed_seq` is just in Python, but would be fine in the Cython world where these are defined?

**Comment by seberg** [comment]
Fair, but I think the right thing is likely still to just get started and see what is actually slow.  If it is the SeedSequence _and_ the SeedSequence is unavoidably slow then one can consider more.

> but would be fine in the Cython world where these are defined?

Yes, you can side-step everything there except `__cinit__` (and I don't think they have a cinit).

**Comment by ricardoV94** [comment]
> IIRC I was initially wondering if we should have something more visible than a deepcopy, implementing __deepcopy__ to make it faster isn't an API issue though and can just be done.

Yes, I may have polluted this issue. Feel free to hide my comments / move them to another issue

**Comment by flying-sheep** [comment]
How should one handle extensibility?

My idea was that in adopting SPEC 7, we make our APIs future proof by making sure
1. each method call always advances the passed rng by exactly the same amount
2. we use `spawn` to pass sub-generators to each subtask, so we can add subtasks without affecting subsequent calls e.g.

   ```py
   rng_init, rng_svds = np.random.default_rng(rng).spawn(2)  # could increase to 3 without changing anything else
   v0 = rng_init.uniform(size=np.min(x.shape))
   u, s, v = svds(..., v0=random_init, rng=rng_svds)
   ```

but this means that when users pass a `Generator`, its state isn’t advanced at all (since `spawn` doesn’t do that).

I could of course add a helper like this, but I feel like if that’s how things should be done, it would be part of the API, right?

```py
def gens_from_root(rng: SeedLike | RngLike | None, n: int) -> list[Generator]:
    rng = np.random.default_rng(rng)
    rng.bit_generator.advance(1)  # advance state of bit generator once
    return rng.spawn(n)  # get independent sub-rngs without advancing the state

def my_pca(x: np.ndarray, rng: SeedLike | RngLike | None = None) -> ...:
    rng_init, rng_b = gens_from_root(rng)
    
    v0 = rng_init.uniform(size=np.min(x.shape))
    u, s, v = svds(x, ..., v0=random_init, rng=rng_svds)
    ...

assert not np.array_equal(my_pca(a), my_pca(a)), "using independent RNGs by passing None should produce different results"
rng = np.random.default_rng()
assert not np.array_equal(my_pca(a, rng=rng), my_pca(a, rng=rng)), "re-using an RNG should advances its state"
np.testing.assert_array_equal(my_pca(a, rng=42), my_pca(a, rng=42), err_msg="using the same seed should produce the same results")
```

**Comment by rkern** [comment]
I'm not sure what you mean by "extensibility". I'm not sure what purpose `gens_from_root()` is serving here. There's no point in `advance()`ing, I don't think. If you are spawning, then the state of the `SeedSequence` is what matters, and it is advancing.

But no, I don't recommend spawning as a regular practice unless if there is actual concurrency going on, if the sub-tasks reorder themselves in some way. They typically don't (and in the contexts where you do, like PyTensor above, you need to write your utility functions that meet the needs of your local patterns).

**Comment by flying-sheep** [comment]
> If you are spawning, then the state of the SeedSequence is what matters, and it is advancing.

The spawned child generators are independent, but the parent generator doesn’t advance:

```pycon
>>> a, b = [np.random.default_rng(42) for _ in range(2)]
>>> _ = b.spawn(1)  # doesn’t advance `b`:
>>> [a.integers(9999), b.integers(9999)]
[np.int64(892), np.int64(892)]
```

The idea was that a function that does nothing but spawning wouldn’t advance the generator if I don’t call `.advance` here.

> if the sub-tasks reorder themselves in some way. They typically don't

They might not reorder, but new steps are added all the time.

E.g. sklearn just added `rng` support to `scipy.sparse.linalg.eigsh`, and added it to `scipy.sparse.linalg.svds` in 1.15. Each time it made sense for use to pass a `rng` to that new parameter to increase reproducibility.

I guess you’re suggesting that we opportunistically add something like `[rng_svds] = rng.spawn(1)` in these cases so we can keep using `rng` how we’ve used it but add reproducibility for the newly-parameterized API without breaking `rng`?

**Comment by rkern** [comment]
Right, but what's the point of advancing the `BitGenerator` state? It's never used in the function, only the `SeedSequence` state is used. I'm not seeing what the problem is that it's solving. The snippet you showed deliberately started by constructing two identical clones and showed that they are still clones. Even if you advance one of them, they'll still be correlated semi-clones that you shouldn't use together anyways. It would be worse if one did advance but not the other because then they would still be correlated with each other, but less detectably. The original error here was constructing two clones and trying to use them together. If you don't do that, then there is no problem. If all the function does with a passed-in `Generator` is spawn it, it's fine to do so. The spawning advances the `SeedSequence` internal state.

Yeah, if you really need to make certain steps reproducible independent of others, sure, spawn. Also consider if they should just be independently seeded to begin with. For example, in ML, I'll always seed my dataset preparation (train/test/val splits) separately from whatever pseudorandomness I use during the training. In that case, it's almost always better to separately seed rather than seed once and spawn. I'll want to change the seeds for training to evaluate robustness, but I never want to redo the dataset preparation.

I don't recommend spawning for each subtask as a widespread pattern, though, but one used occasionally and judiciously. The independent reproducibility is sometimes useful, but is often not particularly necessary compared to the degraded safety of spawning. I don't know why the adoption of SPEC 7 for `svds` and `eigsh` is causing you to think that spawning is needed or desired to obtain the same results as before. SPEC 7 is just about the way that we are moving from `RandomState` to `Generator`. `RandomState` never had spawning; you can get the same reproducibility results with the `rng` argument by very similar patterns as one did with the `random_state` argument. `default_rng()` was designed with similar patterns as scipy's `check_random_state()`.

**Comment by flying-sheep** [comment]
Thanks, that helps! I didn’t know that there are two independent parts of a Generator’s state that can be advanced. I should have tried this:

```pycon
>>> rng = np.random.default_rng(42)
>>> [rng.spawn(1)[0].integers(99999) for _ in range(2)]
[np.int64(49739), np.int64(7469)]
```

> compared to the degraded safety of spawning

I don’t know anything about that, is that explained somewhere?

> SPEC 7 is just about the way that we are moving from `RandomState` to `Generator` […]

Ah sorry, `svds` was a bad example; only `eigsh` just straight-up gained the ability to specify a random seed/state which [it didn’t have in 1.16](https://docs.scipy.org/doc/scipy-1.16.2/reference/generated/scipy.sparse.linalg.eigsh.html).

So it’s a perfect example why one would want to spawn for reproducibility, right? initial code:

```py
# advance `rng` SeedSequence by exactly 1 no matter how much the code changes
[rng_root] = rng.spawn(1)
random_init = rng_root.uniform(...)
eigsh(A, k, v0=random_init)
```

then for scipy 1.17

```diff
 [rng_root] = rng.spawn(1)
+[rng_eigsh] = rng_root.spawn(1)
 random_init = rng_root.uniform(...)
-eigsh(A, k, v0=random_init)
+eigsh(A, k, v0=random_init, rng=rng_eigsh)
```

and then when the next change happens:

```diff
 [rng_root] = rng.spawn(1)
-[rng_eigsh] = rng_root.spawn(1)
+rng_eigsh, rng_other_stuff = rng_root.spawn(2)
 random_init = rng_root.uniform(...)
 eigsh(A, k, v0=random_init, rng=rng_eigsh)
+do_thing(rng_other_stuff)
```

**Comment by rkern** [comment]
I've talked about it earlier in this issue with links. E.g. [I talk about the JAX key splitting at _length_ here](https://github.com/data-apis/array-api/issues/431#issuecomment-1147804063) and [restate the claim more briefly here](https://github.com/data-apis/array-api/issues/431#issuecomment-1150044604), though it's backed by the analysis in the lengthy comment.

In this case, I would just recommend not spawning at all. It's just extra complexity for no real gain. Just pass the one `rng` all the way through. Your end result is going to differ anyways with these code changes; there isn't much point in making the intermediates more stable. All the results are equally valid.

**Comment by flying-sheep** [comment]
I guess since `Generator` no longer makes stability guarantees, we can inherit that instead of adding code to try and stay compatible between versions.

**Comment by ilayn** [comment]
I don't have any comments on the PRNG parts but regarding the svds and eigsh, 

Previously SciPy code was not reproducible anyways because ARPACK had its own random number generator which was not seedable (probably the most common request in https://github.com/opencollab/arpack-ng) Now, being rewritten in C, both PROPACK and ARPACK are seedable externally. So you don't need to care much about reproducibility with the new seeds you might choose. The old usage was working due to the stability of ARPACK and PROPACK and your specific choice of the starting vector.

If the problems you have been testing did not force an implicit restart there would be no change needed. Otherwise you would have seen the values dancing around some fixed value due to internal seeding.

**Comment by charris** [comment]
I think we can close this. Thoughts?
